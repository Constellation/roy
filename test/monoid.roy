typeclass Monoid #a {
  append: Function(#a, #a, #a)
  empty: #a
}

instance stringMonoid = Monoid String {
  append: \x y -> x ++ y
  empty: ""
}

instance productMonoid = Monoid {product: Number} {
  append: \x y -> {product: x.product * y.product}
  empty: {product: 1}
}

instance sumMonoid = Monoid {sum: Number} {
  append: \x y -> {sum: x.sum + y.sum}
  empty: {sum: 0}
}

console.log (append (append "Hello!" empty) empty)

let f x = append empty x
let g x = f x

console.log (g "TEST")

console.log (append {product: 10} (append {product: 2} empty)).product

console.log (append {sum: 1} empty).sum
