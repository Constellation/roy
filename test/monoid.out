Hello!
TEST
20
1
